primitives/trie/src/cache/shared_cache.rs

```diff
@@ -34,94 +33,300 @@ lazy_static::lazy_static! {
-	pub fn get(&self, key: &H) -> Option<&NodeOwned<H>> {
-		self.lru.peek(key)
-	}
+	/// Update the cache with the `list` of nodes which were either newly added or accessed.
+	pub fn update(&mut self, list: impl IntoIterator<Item = (H, NodeCached<H>)>) {
+		let mut access_count = 0;
+		let mut add_count = 0;
 
-	/// Update the cache with the `added` nodes and the `accessed` nodes.
-	///
-	/// The `added` nodes are the ones that have been collected by doing operations on the trie and
-	/// now should be stored in the shared cache. The `accessed` nodes are only referenced by hash
-	/// and represent the nodes that were retrieved from this shared cache through [`Self::get`].
-	/// These `accessed` nodes are being put to the front of the internal [`LruCache`] like the
-	/// `added` ones.
-	///
-	/// After the internal [`LruCache`] was updated, it is ensured that the internal [`LruCache`] is
-	/// inside its bounds ([`Self::maximum_size_in_bytes`]).
-	pub fn update(
-		&mut self,
-		added: impl IntoIterator<Item = (H, NodeOwned<H>)>,
-		accessed: impl IntoIterator<Item = H>,
-	) {
-		let update_size_in_bytes = |size_in_bytes: &mut usize, key: &H, node: &NodeOwned<H>| {
-			if let Some(new_size_in_bytes) =
-				size_in_bytes.checked_sub(key.as_ref().len() + node.size_in_bytes())
-			{
-				*size_in_bytes = new_size_in_bytes;
-			} else {
-				*size_in_bytes = 0;
-				tracing::error!(target: LOG_TARGET, "`SharedNodeCache` underflow detected!",);
-			}
-		};
+		self.lru.limiter_mut().items_evicted = 0;
+		self.lru.limiter_mut().max_items_evicted =
+			self.lru.len() * 100 / super::SHARED_NODE_CACHE_MAX_REPLACE_PERCENT;
 
-		accessed.into_iter().for_each(|key| {
-			// Access every node in the lru to put it to the front.
-			self.lru.get(&key);
-		});
-		added.into_iter().for_each(|(key, node)| {
-			self.size_in_bytes += key.as_ref().len() + node.size_in_bytes();
+		for (key, cached_node) in list {
+			if cached_node.is_from_shared_cache {
+				if self.lru.get(&key).is_some() {
+					access_count += 1;
 
-			if let Some((r_key, r_node)) = self.lru.push(key, node) {
-				update_size_in_bytes(&mut self.size_in_bytes, &r_key, &r_node);
-			}
+					if access_count >= super::SHARED_NODE_CACHE_MAX_PROMOTED_KEYS {
+						// Stop when we've promoted a large enough number of items.
+						break
+					}
 
-			// Directly ensure that we respect the maximum size. By doing it directly here we ensure
-			// that the internal map of the [`LruCache`] doesn't grow too much.
-			while self.maximum_cache_size.exceeds(self.size_in_bytes) {
-				// This should always be `Some(_)`, otherwise something is wrong!
-				if let Some((key, node)) = self.lru.pop_lru() {
-					update_size_in_bytes(&mut self.size_in_bytes, &key, &node);
+					continue
 				}
 			}
-		});
+
+			self.lru.insert(key, cached_node.node);
+			add_count += 1;
+
+			if self.lru.limiter().items_evicted > self.lru.limiter().max_items_evicted {
```

Comment:
Currently this doesn't include replaced nodes, shouldn't we also include them in `items_evicted`?

Reply:
We could then also include the items which we just promote to the head of the lru through `self.lru.get(&key).is_some()` and then just have one number on when to stop instead of also having these `SHARED_NODE_CACHE_MAX_PROMOTED_KEYS`.

Reply:
> Currently this doesn't include replaced nodes, shouldn't we also include them in `items_evicted`?

Hmm... we could, but I'm not sure if there's a point?

First, the main idea of this limit is to make it so that a large local cache can't just wholesale replace the whole shared cache so that generally useful keys don't get evicted from the shared cache just because we've iterated over a big subtree. And a replacement doesn't actually evict anything; it's essentially equivalent to just bumping the entry in the LRU (because the key and the value's going to be the same), just less efficient.

Second, normally there shouldn't actually be any replaced nodes here. Notice that we first check that `cached_node.is_from_shared_cache` and whether `self.lru.get(&key).is_some()`, so *normally* if an item already exists in the shared cache it'll remember that and won't try to `insert` it again, so nothing will be replaced. You'd need at least two local caches running in parallel, both fetching a key which doesn't exist in the shared cache and then getting dropped. This can happen but I think it should be relatively unlikely.

(I think there might have been some confusion here regarding what "replaced" means, considering that you've also asked whether the keys are the same or not in the other comment. :P It's more of a "value_replaced" kind of a thing; if there's no space in the LRU and an item is evicted while you want to insert a new one with a different key then it is *not* a replacement and the `on_replace` callback is *not* called. In that case `on_removed` and `on_insert` is called. I'll rename the callback in the limiter and maybe add a clarifying comment in the next version of `schnellru` to make this more clear.)